src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1318 @@
+      ret = mongoc_stream_poll (poller, 2, (int32_t) timeleft_ms);
+      if (ret == 0) {
+         MONITOR_LOG (server_monitor, "mongoc_stream_poll timeout or error");
+         /* Timeout or error. */
+         bson_set_error (error,
+                         MONGOC_ERROR_STREAM,
+                         MONGOC_ERROR_STREAM_SOCKET,
+                         "poll timeout");
+         return false;
+      } else if (ret == -1) {
+         MONITOR_LOG (server_monitor, "mongoc_stream_poll error");
+         bson_set_error (error,
+                         MONGOC_ERROR_STREAM,
+                         MONGOC_ERROR_STREAM_SOCKET,
+                         "poll error");
+         return false;
+      }
+
+      if (poller[0].revents & (POLLERR | POLLHUP)) {
+         MONITOR_LOG (server_monitor, "mongoc_stream_poll stream error");
+         bson_set_error (error,
+                         MONGOC_ERROR_STREAM,
+                         MONGOC_ERROR_STREAM_SOCKET,
+                         "poll stream error");
+         return false;
+      }
+
+      if (poller[1].revents & (POLLERR | POLLHUP)) {
+         MONITOR_LOG (server_monitor,
+                      "mongoc_stream_poll interrupt stream error");
+         bson_set_error (error,
+                         MONGOC_ERROR_STREAM,
+                         MONGOC_ERROR_STREAM_SOCKET,
+                         "poll interrupt error");
+         return false;
+      }
+
+      if (poller[1].revents & POLLIN) {
+         MONITOR_LOG (server_monitor, "interrupt detected");
+         /* Interrupt. No need to lock here. Interrupt has already been created
+          * and won't be destroyed until shutdown. The interrupt itself is
+          * thread-safe. */
+         _mongoc_interrupt_flush (server_monitor->shared.interrupt);
+         bson_mutex_lock (&server_monitor->shared.mutex);
+         *cancelled = server_monitor->shared.cancel_requested;
+         server_monitor->shared.cancel_requested = false;
+         bson_mutex_unlock (&server_monitor->shared.mutex);
+
+         if (*cancelled) {
+            MONITOR_LOG (server_monitor, "and cancelled");
+            /* Do not set error. */
+            return false;
+         }
+         MONITOR_LOG (server_monitor, "but not cancelled");
+         /* Interrupt triggered, but no cancellation. Try again. */
+         continue;
+      }
+
+      if (poller[0].revents & POLLIN) {
+         MONITOR_LOG (server_monitor, "mongoc_stream_poll ready to read");
+         return true;
+      }
+
+      bson_set_error (error,
+                      MONGOC_ERROR_STREAM,
+                      MONGOC_ERROR_STREAM_SOCKET,
+                      "unexpected error polling streams");
+      return false;
+   }
+   bson_set_error (error,
+                   MONGOC_ERROR_STREAM,
+                   MONGOC_ERROR_STREAM_SOCKET,
+                   "timeout polling");
+   MONITOR_LOG (
+      server_monitor, "timeout polling: exceeded %" PRIu64, expire_at_ms);
```

Comment:
Consider making this a MONGOC_ERROR log, and anything else that is error-worthy.